api/v1alpha1/limitador_types.go

```diff
@@ -102,6 +107,93 @@ type LimitadorList struct {
 	Items           []Limitador `json:"items"`
 }
 
+// StorageType defines the valid options for storage
+// +kubebuilder:validation:Enum=memory;redis;redis_cached
+type StorageType string
+
+const (
+	StorageTypeInMemory    StorageType = "memory"
+	StorageTypeRedis                   = "redis"
+	StorageTypeRedisCached             = "redis_cached"
+)
+
+// Storage contains the options for Limitador counters database or in-memory data storage
+type Storage struct {
+	// +optional
+	Redis *Redis `json:"redis,omitempty"`
+
+	// +optional
+	RedisCached *RedisCached `json:"redis-cached,omitempty"`
+}
+
+func (s *Storage) Validate() bool {
+	return s.Redis != nil && s.Redis.ConfigSecretRef != nil ||
+		s.RedisCached != nil && s.RedisCached.ConfigSecretRef != nil
+}
+
+func (s *Storage) SecretRef() *corev1.ObjectReference {
+	if s.Redis != nil {
+		return s.Redis.ConfigSecretRef
+	}
+	return s.RedisCached.ConfigSecretRef
+}
+
+func (s *Storage) Config(url string) []string {
+	if s.Redis != nil {
+		return []string{
+			StorageTypeRedis,
+			url,
+		}
+	} else if s.RedisCached != nil {
+		return []string{
+			StorageTypeRedisCached,
+			url,
+			fmt.Sprintf("--ttl %d", s.RedisCached.Options.TTL),
+			fmt.Sprintf("--ratio %d", s.RedisCached.Options.Ratio),
+			fmt.Sprintf("--flush-period %d", s.RedisCached.Options.FlushPeriod),
+			fmt.Sprintf("--max-cached %d", s.RedisCached.Options.MaxCached),
+		}
+	}
+	return []string{string(StorageTypeInMemory)}
+}
+
+type Redis struct {
+	// +ConfigSecretRef refers to the secret holding the URL for Redis.
+	// +optional
+	ConfigSecretRef *corev1.ObjectReference `json:"configSecretRef,omitempty"`
+}
+
+type RedisCachedOptions struct {
+	// +optional
+	// +kubebuilder:default=5000
+	// TTL for cached counters in milliseconds [default: 5000]
+	TTL int `json:"ttl,omitempty"`
+
+	// +optional
+	// +kubebuilder:default=10
+	// Ratio to apply to the TTL from Redis on cached counters [default: 10]
+	Ratio int `json:"ratio,omitempty"`
+
+	// +optional
+	// +kubebuilder:default=1000
+	// FlushPeriod for counters in milliseconds [default: 1000]
+	FlushPeriod int `json:"flush-period,omitempty"`
+
+	// +optional
+	// +kubebuilder:default=10000
+	// MaxCached refers to the maximum amount of counters cached [default: 10000]
+	MaxCached int `json:"max-cached,omitempty"`
+}
+
+type RedisCached struct {
+	// +ConfigSecretRef refers to the secret holding the URL for Redis.
+	// +optional
+	ConfigSecretRef *corev1.ObjectReference `json:"configSecretRef,omitempty"`
+
+	// +optional
+	Options *RedisCachedOptions `json:"options,omitempty"`
```

Comment:
I would save `RedisCachedOptions`. I wonder if all the options should  be at the same level as the config secret ref

Reply:
I place them separated just to make a distinction between arguments (URL) and Options, not sure if that makes sense